nomad/structs/network.go

```diff
@@ -554,10 +557,14 @@ func (idx *NetworkIndex) AssignPorts(ask *NetworkResource) (AllocatedPorts, erro
 			// lower memory usage.
 			var dynPorts []int
 			// TODO: its more efficient to find multiple dynamic ports at once
```

Comment:
Note to reviewers: this comment is really only true in the case where the dynamic range is full enough that we fail to get randomly selected ports. Optimizing for this will involve heavily refactoring this method to figure out the count of dynamic ports we need (per network/address), getting the offered ports, and then dolling them out for each of the asked ports. I want to do this performance work in the very near term, but not in the same PR as a major bug fix.

Reply:
Sounds like a fun problem! I would even go so far as to suggest reconsidering the use of a `Bitmap`, perhaps instead thinking of our own data structure that provides an efficient API surface that actually makes sense here (eliminating the need for random tries with precise fallback). No reason it can't be done, it's just off the beaten path.